File: leetcode-python/medium/_518_coin_change_2/solution2.py
from typing import List


class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        memo = {}

        def dp(i, target):
            if target == 0:
                return 1
            elif i == 0 and target != 0:
                return 0

            if (i, target) in memo:
                return memo[(i, target)]
            result = 0
            coin_num = target // coins[i - 1]
            for j in range(0, coin_num + 1):
                result = result + dp(i - 1, target - j * coins[i - 1])
            memo[(i, target)] = result
            return result

        return dp(len(coins), amount)

File: leetcode-java/src/main/java/medium/_518_coin_change_2/Solution2.java
package medium._518_coin_change_2;

// dp - bottom up solution
public class Solution2 implements ISolution{
    public int change(int amount, int[] coins) {
        int n = coins.length;
        int[][] dp = new int[n + 1][amount + 1];
        for (int i = 0; i <= n; i++)
            dp[i][0] = 1; // have 1 way to change 0 amount
        for (int c = 1; c <= n; c++) {
            for (int a = 1; a <= amount; a++) {
                dp[c][a] = dp[c - 1][a]; // skip ith coin
                if (a >= coins[c - 1])
                    dp[c][a] += dp[c][a - coins[c - 1]]; // use ith coin
            }
        }

        return dp[n][amount];
    }

}
